Return & use error code from OpenVR initialisation

shu8 opened the pull request:
This is part of https://github.com/webaverse/metachromium-bin/issues/12.

Previously, when the OpenVR initialisation failed, we didn't do anything with the error code.

This commit changes the behaviour of the qrvr C++ module to return the error code received from OpenVR, and the JS will exit the Node process with the returned error code, if non-0.

I've also added some details to the README that explain how to build the package locally, and the dependencies you need (after trying to get it to build myself, with help from @avaer)

shu8 committed:
Return & use error code from OpenVR initialisation

This is part of https://github.com/webaverse/metachromium-bin/issues/12.

Previously, when the OpenVR initialisation failed, we didn't do anything
with the error code.

This commit changes the behaviour of the qrvr C++ module to return the
error code received from OpenVR, and the JS will exit the Node process
with the returned error code, if non-0.

shu8 committed:
Add development documentation

This adds some pretty important development instructions, mainly the
system dependencies needed to successfully build the package!

avaer commented on README.md:
````diff
@@ -1,15 +1,50 @@
 ```
 const qrvr = require('qrvr');
-qrvr.start().then(s => {
+const exitCode = qrvr.start().then(s => {
````
I'm not sure of this will have the exit code. Won't this be a `Promise`?

avaer replied:
I see, so the intent is to capture the exit code. However, this isn't how promises work in ES5.

First, we would probably need to `await` here. Secondly, I don't think the promise itself is resolving to the exit code, or even waiting for the termination of the process (see below).

avaer replied:
If you log `exitCode` I suspect you will see it is a promise that resolves to a string.

shu8 replied:
I actually didn't test that snippet, my bad sorry. I'll try testing it as soon as I can! My thinking was that because I was `process.exit`ing, and because we're not `.catch`ing the `start` promise, it would just end with that status code but I might be wrong so will check! :)

avaer commented on index.js:
```diff
@@ -18,7 +18,18 @@ async function start({
+      // See https://github.com/ValveSoftware/openvr/wiki/HmdError for error codes
+      if (resultCode === 100) {
+        console.error('OpenVR Installation not found', resultCode);
+        return process.exit(resultCode);
```
I don't think this will do what was intended.

It will presumably return the result of `process.exit` to the handler of `presenceWss.on`. I'm not sure why that would be useful -- it will probably just be thrown out in that handler.

The intended thing was probably to return it as the result of some promise in this closure, which would need a different strategy.

shu8 replied:
Hmm it did seem to work when I test it, mainly because I think the `process.exit` quit the entire node process immediately with the `resultCode` status code. The `return` in that line is probably unnecessary though because the exit will stop any more code from running anyway.

But I think you're suggesting wrapping the whole `start` function in a promise and just calling `reject(resultCode)` instead of the `return process.exit(resultCode)` which I could do instead?

avaer replied:
> But I think you're suggesting wrapping the whole start function in a promise and just calling reject(resultCode) instead of the return process.exit(resultCode) which I could do instead?

Yes, that might work.

avaer reviewed (changes requested):
This looks like the right direction but I don't think it's working at the moment.

shu8 committed:
Refactor start method to only resolve on successful QR Engine init

This changes the start method to only resolve the promise if the QR
engine successfully starts up. Previously, it resolved as soon as the
websocket server started but that meant if there was a QR Engine crash,
the promise couldn't be rejected easily.

shu8 commented:
@avaer I've updated the module to only resolve the `start` method if the QR Engine was successfully initialised. Previously, it resolved the promise as soon as the express websocket server started but that meant if there was a QR Engine crash, the promise couldn't be rejected easily.

avaer commented on index.js:
```diff
@@ -1,43 +1,52 @@
       s.on('message', s => {
         const j = JSON.parse(s);
-        const {method} = j;
+        const { method } = j;
```
Is this being done by lint rules?

shu8 replied:
It was! I've added the actual eslint semistandard config now and re-linted the code

avaer replied:
Is there a way to not add these spaces in the linter for this PR?

There might be a time to re-format files, but I don't think it should be done with spacing changes sprinkled into unrelated PR's.

shu8 replied:
Sure, I've removed the spaces

avaer commented on index.js:
```diff
@@ -1,43 +1,52 @@
 const ws = require('ws');
 const engine = require('./build/Release/qr.node');
 
+async function initQr(qrEmitter, locomotionEmitter) {
```
If the function is already `async` then there is no need to also wrap in a `Promise`.

This function looks like it should be `async` and throw rather than returning a `Promise` in the body.

shu8 replied:
I'm not sure how to `throw` just an error code like I am currently doing with the promise rejection.

I could do `throw new Error(resultCode)` and then in the code checking for the specific error in metachromium (https://github.com/webaverse/metachromium/pull/4/files#diff-168726dbe96b3ce427e7fedce31bb0bcR67) do something like `if (error.toString() === 'Error: 100')` instead but it feels a bit more hacky so not sure what the best way to do it is!

avaer replied:
`async` functions will reject the `Promise` as the result of a `throw`.

shu8 replied:
But it will reject the Promise as an actual `Error` object rather than just the `resultCode` I pass into the throw I think?

avaer replied:
Couldn't it `throw` the number?

shu8 replied:
Ah, TIL you don't have to `throw` an actual `Error`.

Thanks! Will push that now :)

shu8 replied:
I just pushed this change!

avaer reviewed (changes requested):
This looks like it should work, just some small code questions.

shu8 committed:
Remove unnecessary Promise inside async function

avaer commented on index.js:
```diff
@@ -1,43 +1,49 @@
+async function start ({
   port = 8000,
-  key,
+  key
```
Trailing commas are very prevalent and preferred in Webaverse code so this rule will probably cause ongoing conflicts.

shu8 replied:
I've removed the eslint config for now

avaer reviewed (commented):
This looks good, though it entangles unrelated formatting with the substance of the PR.

When re-linting projects, can we do that in a separate PR?

shu8 committed:
Revert whitespace changes (for now)

Linting can be done later.

avaer reviewed (approved):
👏

shu8 commented:
Thanks! Do we need to bump the version number before releasing the new version to npm?
